### ISSUE TITLE
Operations that have multiple options

### ISSUE THREAD

**Issue by IanMayo** [debrief/limpet_ops#14]
For non associative operations,  we need to provide UI options to specify the order of the arguments. 

We also need to provide multiple options for some other operations.   One such operation is to resample the data.   For a resample operation,  we offer a choice of frequencies (something like::   1 sec,  5 secs,  10sec, 30 secs,  1 minute, 2 mins,   5 mins,  10 mins,  15 mins,  30 mins,  1 hour,  2 hours, 5 hrs,  12 hours,  1 day). 

Hmm,  do we need to represent this declarative?   We "could"  just handle it at the programmatic stage.   But,  if we handle it declaratively we can generate the menu options without loading the actual .jar file.

Other examples of operations that require multiple options are:
- moving average (with choice of window size)
- apply smooothing (choice of smoothing algorithm)
- apply filter (again, with choice of algorithm)
- [come on Ian, think of some more....]

**Comment by dinkoivanov** [comment]
Yes, we can define parameters as an additional extension object and make it declarative, so each operation defines own set of parameters. Parameters can be defined with their type (number, string, etc), perhaps some range validation and more. So when the user selects an operation, say "Moving average" we can either nest another submenu with some predefined "window size" options, or if there are no predefined options (or there are more params to specify), we can display a popup dialog to configure the parameters. The plugin declaring the operation won't be loaded until user configures all params and presses OK/Execute.
I'm not sure the declarative params approach is a huge benefit against having the operation execute method to create some UI and configure the params (programmatic approach). When user clicks on an operation from the menu, he has given consent that he wants to perform the operation, so I think it's good time to load the plugin at this point (whether or not he has configured the params).
The only benefit of declarative approach would be if we have indeed such list of predefined parameters (like sampling rate, window size) where user will see a context submenu to choose from.

**Comment by IanMayo** [comment]
>  When user clicks on an operation from the menu, he has given consent 
> that he wants to perform the operation, so I think it's good time to load the 
> plugin at this point (whether or not he has configured the params).

I think your describing that once the user has selected _Moving  average_, we dynamically populate the child menu using Java code.  Isn't there a chance that this will introduce a UI delay, as the .jar is loaded?

There may be some operations where it's expensive to determine what to offer in the child menu, so it's right to only perform it once the user has selected the parent menu item (else construction of our menu tree would become slow with lots of these being performed "just in case" the user selected that branch).  But, largely, I prefer the declarative option.

I'm also loosely considering how a web-based instance of Limpet would perform. The declarative menu structure could sit on the client, but each time we require programatic functionality we'll probably require a server round trip.  But, hey - that's in the distant future :-)

**Comment by dinkoivanov** [comment]
Another thing to consider - in the main limpet repo, we have the UIProperty annotation concept. If we go with the programmatic approach I think we'll be able to reuse it - the operation parameters will be defined with some annotation metadata, then the code which constructs the UI for operation library would be able to read the metadata and create some configuration UI for the parameters.
With the declarative approach we'll need to define the parameters metamodel (type, default value, range, etc.) in the extension schema, which is kind of duplicating what's already there in the UIProperty annotation. 
However if the modularity and lazy-loading of plugins is top prio, then this reuse should not be that big deal and we can go with the declarative approach. What do you think?

**Comment by IanMayo** [comment]
Hmm, good thinking. 

Let's stick with the UIProperty annotation.  Could you check if we have an example of an operation with multiple options expressed in the UIProperty? If there isn't an example, could you create one?

**Comment by IanMayo** [comment]
@dinkoivanov - here's a _blast from the past_... 

Could you check if we produced a demo for an operations with multiple permutations?
